api/java/org/openyolo/api/internal/ProviderResolver.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2017 The OpenYOLO Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.openyolo.api.internal;
+
+import static org.openyolo.protocol.ProtocolConstants.OPENYOLO_CATEGORY;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.pm.ResolveInfo;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Utility for producing OpenYOLO resolving the activities exposed by providers to handle specific,
+ * OpenYOLO actions.
+ */
+public final class ProviderResolver {
+
+    /**
+     * Creates an explicit Intent to invoke the specified action on the specified provider,
+     * if the provider has a handler registered.
+     */
+    @Nullable
+    public static Intent createIntentForAction(
+            @NonNull Context applicationContext,
+            @NonNull String providerPackageName,
+            @NonNull String action) {
+
+        ComponentName providerComponent =
+                findProvider(applicationContext, providerPackageName, action);
+
+        if (providerComponent == null) {
+            return null;
+        }
+
+        Intent intent = new Intent(action);
+        intent.setClassName(
+                providerComponent.getPackageName(),
+                providerComponent.getClassName());
+        intent.addCategory(OPENYOLO_CATEGORY);
+        return intent;
+    }
+
+    /**
+     * Resolves a {@link ComponentName} for the activity defined in the specified provider, for
+     * the specified action.
+     */
+    @Nullable
+    public static ComponentName findProvider(
+            @NonNull Context applicationContext,
+            @NonNull String providerPackageName,
+            @NonNull String action) {
+        List<ComponentName> providers = findProviders(applicationContext, action);
+
+        for (ComponentName provider : providers) {
+            if (providerPackageName.equals(provider.getPackageName())) {
+                return provider;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Resolves the {@link ComponentName components} for all providers which can handle the
+     * specified OpenYOLO action.
+     */
+    @NonNull
+    public static List<ComponentName> findProviders(
```

Comment:
There is a some similar logic inside CredentialClient, could this replace that?

Reply:
Yes, it can, but wanted to keep this PR focused on this specific issue. Will refactor other code in follow-up.

Reply:
SGTM, a nice middle ground might be to mark the old methods as deprecated with a reference to this class.